### DISCUSSION TITLE
Complex relationship and JOIN request

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7746] [Usage Questions]
I have this model definition : 

```python
from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.orm import relationship

class Ke(Base):
    __tablename__ = 'ke'
    id = Column(Integer, primary_key=True)


class Va(Base):
    __tablename__ = 'va'
    id = Column(Integer, primary_key=True)
    k_id = Column(Integer, ForeignKey('ke.id'))


class Vi(Base):
    __tablename__ = 'vi'
    id = Column(Integer, primary_key=True)
    va_id = Column(Integer, ForeignKey('va.id'))   
    ge_id = Column(Integer)     

class T1(Base):
    __tablename__ = 't1'
    id = Column(Integer, primary_key=True)
    e_id = Column(Integer)
    k_id = Column(Integer, ForeignKey('ke.id'))
    t2s = relationship('T2')
    vis = relationship("Vi", foreign_keys="[Va.k_id, Vi.ge_id, Vi.va_id, Va.id]",
                             secondary="join(Va, Vi, Vi.va_id == Va.id)",
                             primaryjoin="T1.e_id==Vi.ge_id",
                             secondaryjoin="T1.k_id==Va.k_id",
                             viewonly=True))
```
I now have this query 

```python
query = select(T1).outerjoin(T1.t2s).options(contains_eager(T1.t2s)).outerjoin(T1.vis).options(contains_eager(T1.vis))
```

Which output me this SQL Query 

```sql 
SELECT 
 t2.id
 t1.k_id,
 t1.e_id,
 t1.id AS id_1
FROM t1
JOIN t2 ON t1.id = t2.g_id
JOIN (va as va,
    JOIN vi AS vi_1 ON vi_1.va_id = va.id
    JOIN vi ON t1.k_id = va.k_id) 
ON t1.e_id = vi_1.ge_id;
```
Which raises this SQL error : 

```sql
 Unknown column 't1.k_id' in 'on clause'
````

I understand from this error that basically MySQL resolve first the JOIN within the parenthesis, and t1 is not recognized since it is not declared yet. So the query does not render correct SQL syntax (I use mysql dialect)

I am not sure if I used the correct syntax in the complex relationship called `vis` on `T1`, did I? 

I did not put the T2 definition but it is a simple table with a foreign key to T1

**Comment by zzzeek** [comment]
the secondary join reaching out to a column that's not actually in the target table suggests you should use the [relationship to aliased class](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#relationship-to-aliased-class) pattern.   that will likely give you the additional flexibility to keep the target columns surfaced at the top of the selectable so that you can target them in the top-level relationship join.
